### DISCUSSION TITLE
custom (ESM) loader + extension-less "bin" files

### DISCUSSION THREAD

**Discussion by JakobJingleheimer** [nodejs/help#4904] [Q&A]
Currently, when a custom ESM loader is used in combination with an extension-less commonjs file, it erroneously fails. For example

```
project-a/
  ├ node_modules/
     └ mocha/
        ├ bin/
           └ mocha
        └ package.json
  ├ src/
  ├ a-loader.cjs
  ├ a-loader.mjs
  └ package.json
```

```
NODE_OPTIONS="--loader=./a-loader.mjs" npx mocha ./src
```

There have been several disparate discussions about this issue. I’ve tried to capture the main points here from the ones I’ve found so far. If I’ve missed any, please comment and I’ll update.

Related: #33226, #41275, #41465, #41552, mochajs/mocha#4645

# Problems

**1A**: The presence of the `--loader` flag [causes Node.js to use the "wrong" loader](https://github.com/nodejs/node/blob/ccb8aae3932c13f33622203b2ffc5a33120e9d40/lib/internal/modules/run_main.js#L30-L45) (ESMLoader instead of CJSLoader), regardless of the entry-point’s actual type.

**1B**: [`ESMLoader::defaultGetFormat()`](https://github.com/nodejs/node/blob/ccb8aae3932c13f33622203b2ffc5a33120e9d40/lib/internal/modules/esm/get_format.js#L75-L90) ignores any file without a `.cjs`, `.mjs`, `.js`, `.json`, or `.node` file extension. This means, for a commonjs file without a file extension (which is allowed), Node.js will eventually throw when it should not.

# Considerations

## WASM

ESMLoader intentionally does not handle extension-less files due to the possibility of extension-less wasm files in future.

## Loader for entry-point

There is not an authoritative way to dictate which Loader (CJS vs ESM) should be used: A user specifying `--loader` my intend for that to be used when loading the entry-point (and thus manipulate how it’s handled). However, the logic in `shouldUseESMLoader()` does not account for CJS’s equivalent flag, `--require` (and friends). So if both are supplied together:

```
NODE_OPTIONS="--loader=./a-loader.mjs --require=./a-loader.cjs" npx mocha ./src
```

What does that mean? One might reasonably expect the ESM loader to merely be used when the parent is ESM and the CJS require to merely be used when the parent is CJS (there is currently no other way for the ESM loader) and neither to affect which Loader is used for the entry-point.

# Previously proposed solutions

I've created threads below with each I found.

# Workarounds

## Add a snippet to the custom loader

I use this in my own projects (my actual uses a map of extensions to load “hooklets” à la `{ '': loadBin, jsx: loadJSX, … }`)

<details>
<summary>sample</summary>

```mjs
export async function load(resolvedUrl, context, defaultLoad) {
  const url = new URL(resolvedUrl);
  const ext = path.extname(url.pathname).slice(1);

  if (ext === '') loadBin(url, context, defaultLoad);

  // … other custom loader code

  return defaultLoad(resolvedUrl, context);
}

async function loadBin(url, context, defaultLoad) {
  const dirs =  path
    .dirname(url.pathname)
    .split(path.sep);
  const parentDir = dirs.at(-1);
  const grandparentDir = dirs.at(-3);

  let format;

  if (parentDir === 'bin' && grandparentDir === 'node_modules') {
    const libPkgUrl = new URL('../package.json', url)
    const { type } = await fs.readFile(libPkgUrl).then(JSON.parse);

    format = type === 'module' ? 'module' : 'commonjs';
  }

  return defaultLoad(url.href, {
    ...context,
    format,
  });
}
```
</details>

@nodejs/loaders

**Comment by cspotcode** [comment]
> 1A: The presence of the --loader flag causes Node.js to use the "wrong" loader (ESMLoader instead of CJSLoader), regardless of the entry-point’s actual type.

Another way to phrase this might be that the ESMLoader is the correct loader in all cases, but it arrives at a different / "wrong" conclusion compared to the CJSLoader.  Both loaders should come to the same conclusion regarding the format of the entrypoint.

Teaching the ESMLoader today to draw the same conclusion as the CJSLoader for entrypoints does not prevent node from, in the future, updating *both* loaders to arrive at a different conclusion.  But leaving node's two loader behaviors inconsistent as it is today makes this a problem.

**Reply by bmeck** [reply]
The ESMLoader cannot come to the same conclusion as the CJSLoader in a forwards compatible way. The CJS loader treats all unknown file extensions (including extension-less) as if it were a `.js` file. Having it come the same conclusion as CJS currently comes to prevents the possibility of ESM in an extension-less file for example or introducing a new importable file extension (such as `.wasm` potentially).

**Reply by bmeck** [reply]
I'm not disagreeing that ESM as the basic entrypoint always is potentially the correct path, but how to solve things in a backwards compatible and forwards compatible way is hard.

**Reply by JakobJingleheimer** [reply]
>> 1A: The presence of the --loader flag causes Node.js to use the "wrong" loader (ESMLoader instead of CJSLoader), regardless of the entry-point’s actual type.
>
> Another way to phrase this might be that the ESMLoader is the correct loader in all cases, but it arrives at a different / "wrong" conclusion compared to the CJSLoader.

I think these two statements are not equivalent (although they could eventually both be true). For reasons Bradley mentioned, they can't always come to the same conclusion, but they can come to similar conclusions (but doing so now makes some wiser people wary).

**Reply by guybedford** [reply]
Could we perhaps reserve a list of future file extensions, such as `wasm`, `wat`, `wit`, etc to create a common subset where the loaders can converge?

**Reply by bmeck** [reply]
@guybedford to some extent yes, but likely we still need to have one win out in the end, given that CJS accepts anything it likely would be the one and regardless of if they converge it would still need to be forwards incompatible for backwards compatibility unless we introduce a breaking change to make it forwards compat.

**Reply by cspotcode** [reply]
To clarify, my proposal here is a deliberately small and forwards-incompatible change that I believe is the correct change to implement today.  It seems there will be other breaking changes regardless of what we decide here, for example, defaulting extensionless entrypoints to ESM.

This change lets us use the ESM loader by default today, with or without `--loader`.  And the only undesirable aspect is that the ESMLoader will behave *for entrypoints* not exactly how we want it to in the future.  Non-entrypoint behavior will be unaffected and forwards-compatible.

#### Pros:
- loaders can continue to operate on entrypoints, which is more useful and less complex than making them an exception to the rule
- ESM loader can continue to throw errors for extensionless non-entrypoints.  (good, forwards-compatible)

#### Cons:
- ESM loader's `getFormat` will have a future breaking change *only for extensionless entrypoints*.  All other behaviors are forward-compatible.

We might also consider raising a warning for extensionless entrypoints that pass through the loader.  This lets us warn the ecosystem about future changes without outright breaking them today.

**Reply by bmeck** [reply]
> defaulting extensionless entrypoints to ESM.

This is not a goal / unlikely to ever happen due to sheer amount of breakage so we can move discussion of it elsewhere.

To note, using the ESM loader can still load CJS files if it knows to do so so using the ESM loader for the entry point is very different from treating the entry point as ESM (terms here are quite confusing unfortunately).

> loaders can continue to operate on entrypoints, which is more useful and less complex than making them an exception to the rule

This is complex an an exception that causes various other workflows to fail which is why this approach keeps failing to gather consensus.

In particular it means that:

```console
node bin/foo
```

Would go through the ESM loader defaults but cannot be wrapped via node properly since

```console
node -e "import('bin/foo')"
```

Would fail since it is not an entrypoint. This kind of workflow of wrapping an entrypoint is used in the wild and doesn't have a workaround while adding an extension to the entrypoint is a simple workaround to problems with extension-less files.

> ESM loader's getFormat will have a future breaking change only for extensionless entrypoints. 

This seems fine as long as we know what that change actually is, but as it stands this seems a bit confusing to me. I'd rather have a solution that can be migrated iteratively towards rather 2 potentially different breaking changes over time with a warning that the solution for this issue shouldn't be considered stable.

**Reply by guybedford** [reply]
```
node -e "import('bin/foo')"
```
> Would fail since it is not an entrypoint. This kind of workflow of wrapping an entrypoint is used in the wild and doesn't have a workaround while adding an extension to the entrypoint is a simple workaround to problems with extension-less files.

In the above, using a child_process spawn instead would still ensure the workflow is supported. That is - there is still a sensible way to achieve this, while solving the major use case.

So I don't think that is a compelling argument against this approach.

**Reply by cspotcode** [reply]
> This is not a goal / unlikely to ever happen due to sheer amount of breakage so we can move discussion of it elsewhere.

Agreed, I was just trying to play along.

> To note, using the ESM loader can still load CJS files if it knows to do so so using the ESM loader for the entry point is very different from treating the entry point as ESM (terms here are quite confusing unfortunately).

Correct, in everything I've said about using the ESM loader for entrypoints, I am referring to invoking ESM loader getFormat.  I am *not* saying that the entrypoint is necessarily executed as ESM.  The output of `getFormat` should be respected, whatever that may be, wherever it may come from (external loader or node's internal loader / `default*` functions)

**Reply by bmeck** [reply]
child_process.spawn causes things like PID tracking to need to change / is complex regarding process.execArgv / loses global mutations / etc. I agree it is a mitigation but not a simple or easy one. This doesn't just affect process entry points as well, this mitigation also affects Workers which have similar complexities vs adding a file extension.

**Reply by guybedford** [reply]
Yes, but if we are able to let users write extensionless Node.js applications that can support loader hooks and ESM that is still a win.

**Reply by bmeck** [reply]
@guybedford we are in agreement on being able to use loader hooks with extension-less files being a good thing, but the solution isn't without downsides or mitigations to those downsides. We are just trying to figure out a solution on how to handle that in this discussion. I think having exceptions based upon way in which something is imported is complex. Additionally, even without the concerns above doing this type of exception cannot be statically determined if it relies on something being an entry point. This approach just has a lot of complexity that spills out vs other ways to allow loaders to handle extension-less files.

**Comment by JakobJingleheimer** [comment]
# Proposal: Limited extension-less “bin” support in `defaultGetFormat()`

Rationale: It is simple to determine when the situation is occurring, and similarly simple to address here.

This proposal was all but rejected.

**Reply by bmeck** [reply]
I'm wondering if this should be revisited but not as an exception based upon entry point / extension-less. Perhaps having ESM loader default return the format from/as CJS if in a type:commonjs context would fix most issues/all? It has a major downside of any new extension that ESM picks up it wouldn't go through the CJS loader anymore which is a forwards compatibility hazard, but only a hazard in type:commonjs contexts.

**Reply by JakobJingleheimer** [reply]
I don't quite understand what you're suggestion. Are you saying `ESMLoader::defaultGetFormat()` should pass anything it doesn't understand to CJSLoader (where it may eventually error)?

**Reply by bmeck** [reply]
Whenever an unknown format is found in a type:commonjs context, yes. Since resolve() can return a hinted type this actually works pretty well.

```
/a #type:module
/a/foo.mts

/a/bin #type:commonjs
/a/bin/bar
```

```console
# no --loader
node /a/foo #in type:module
# esm resolve('file:///a/foo', {parentUrl: undefined}) -> {url: 'file:///a/foo', format: undefined}
# esm load(...) // ERROR: unknown format + no resolved format

node --loader mts /a/foo.mts #in type:module
# esm CUSTOM resolve('file:///a/foo.mts', {parentUrl: undefined}) 
  # esm resolve -> {url: 'file:///a/foo.mjs', format: undefined}
# -> {url: 'file:///a/foo.mjs', format: 'typescript'}
# esm CUSTOM load(...)
  # esm load(...) // unknown format but has resolved format, return source but type:'typescript'
# -> {...}

node /a/bin/bar #in type:commonjs
# esm resolve('file:///a/bin/bar', {parentUrl: undefined}) -> {url: 'file:///a/bin/bar', format: undefined}
# esm load(...) // unknown format + no resolved format
# -> {..., format:'commonjs'}

node --loader mts /a/bin/bar #in type:commonjs
# esm CUSTOM resolve('file:///a/bin/bar', {parentUrl: undefined})
    # esm resolve -> {url: 'file:///a/bin/bar', format: undefined}
# esm CUSTOM load(...)
  # esm load(...) // unknown format + no resolved format -> {..., format:'commonjs'}
```

**Comment by JakobJingleheimer** [comment]
# Proposal: Remove flag checks from `shouldUseESMLoader()`

Rationale: The current inclusion of the flag checks is undocumented behaviour and violates documented behaviour for determining how a file is loaded. Consideration of this other issue is larger than the bug at hand.

This proposal seemed to have a fair number of supporters, but also some opponents. Opposition cited the (very small) burden it could put on certain users, despite [very simple and very viable workarounds](https://github.com/nodejs/node/pull/41552#issuecomment-1018011463).

**Reply by GeoffreyBooth** [reply]
If the problem we’re trying to solve is “Currently, when a custom ESM loader is used in combination with an extension-less commonjs file, it erroneously fails,” as the top post says, then this addresses the issue; and it fixes Node to follow expected behavior. We should do this.

Doing only this wouldn’t provide a way to use an extensionless _ESM_ file, but that strikes me as a less urgent (and separate) problem. The workarounds for that are very simple (a one-line extensionless CommonJS file, for example).

**Reply by cspotcode** [reply]
Does it allow doing `node --loader tsloader x.mts`?  Where x.mts is transformed in-memory and executed as ESM?

**Reply by GeoffreyBooth** [reply]
> Does it allow doing `node --loader tsloader x.mts`? Where x.mts is transformed in-memory and executed as ESM?

That’s not the problem we’re trying to solve.

**Reply by cspotcode** [reply]
Yes, agreed, I'm double-checking that it doesn't introduce another problem, asking a question as a way to confirm my understanding of the proposal.

**Reply by JakobJingleheimer** [reply]
`node --loader tsloader x.mts` does currently work due to this bug. I definitely think that should be properly supported, but leaving it to work by depending on a bug is not a proper solution.

A simple workaround whilst we figure out proper support (which affects more than just TypeScript) is:

```console
node --loader=loader.mjs --input-type=module -e "import 'test.mts'"
```

<details>
<summary>which I confirmed does work</summary>

```mjs
import path from 'path';

import esbuild from 'esbuild';


export async function load(resolvedUrl, context, defaultLoad) {
  const url = new URL(resolvedUrl);
  const ext = path.extname(url.pathname).slice(1);

  const hooklet = hookletExtMap[ext];

  if (hooklet) return hooklet(url, context, defaultLoad);

  return defaultLoad(resolvedUrl, context);
}

const hookletExtMap = {
  '': loadNaked,
  'mts': loadTypeScript,
  'ts': loadTypeScript,
};

async function loadNaked(url, context, defaultLoad) {
  const dirs =  path
    .dirname(url.pathname)
    .split(path.sep);
  const parentDir = dirs.at(-1);
  const grandparentDir = dirs.at(-3);

  let format;

  if (parentDir === 'bin' && grandparentDir === 'node_modules') {
    const libPkgUrl = new URL('../package.json', url)
    const { type = 'commonjs' } = await fs.readFile(libPkgUrl).then(JSON.parse);

    format = type;
  }

  return defaultLoad(url.href, {
    ...context,
    format,
  });
}

async function loadTypeScript(url, context, defaultLoad) {
  const format = 'module';
  const rawSource = '' + (await defaultLoad(url.href, { format })).source;

  const { code: source } = await esbuild.transform(rawSource, {
    format: 'esm',
    loader: 'ts',
    sourcemap: 'inline',
    target: 'esnext',
  });

  return {
    format,
    source,
  };
}
```

```ts
const foo: number = 1;

export default foo;
```
</details>

**Reply by guybedford** [reply]
> node --loader tsloader x.mts does currently work due to this bug.

@JakobJingleheimer which bug are you referring to here?

**Reply by JakobJingleheimer** [reply]
ESMLoader being used instead of CJSLoader for a CJS entry-point, and then failing.

**Reply by guybedford** [reply]
Right, but there is a correction space because you can use the right loader hooks to make it work as necessary. Whereas the converse for the CJS loader isn't possible.

**Reply by cspotcode** [reply]
It would be better to skip the ESM loader for extensionless entrypoints but not for extension-ed entrypoints.  That allows loaders to continue operating on entrypoints and fixes the extensionless bug.  It is a smaller, tactical change to behavior.

Note that I'm not saying this behavior needs to be permanent; certainly breaking changes can and will be made in the future.  But this is a correct solution for today, allowing the ecosystem to continue using `--loader` and give meaningful feedback without imposing unnecessary breakages which distract from `--loader`'s intended purpose.  I believe this is what we want.

**Reply by bmeck** [reply]
Is there a desire/use case for allowing extension-less entrypoints in type:module ? I'd rather agree that using the ESM loader always for type:module rather than based upon flags seems a decent compromise

**Reply by cspotcode** [reply]
With `type:commonjs` I think loaders / users will want to execute *extensioned* entrypoints as ESM: `.mts` maps to `.mjs`, so an `.mts` entrypoint should be ESM, ignoring package.json type.  I think this is clean and intuitive and maps nicely to what node already does for `.mjs`

But I can't think of a good reason why a loader would want to make an *extensionless* entrypoint execute as ESM.

**Reply by GeoffreyBooth** [reply]
I think these are two problems. First, flags should have no bearing on how an entry point is resolved, full stop. (Unless/until we create a flag for that purpose.) That’s just a bug that needs to be fixed. Yes it would break `node --loader tsloader x.mts`, but like @JakobJingleheimer said, the fact that that works at all is an unintended side effect of this bug.

The second problem is how to allow users to use an `.mts` file as an entry point. That might warrant its own discussion thread.

One option is to just always use the ESM loader for entry points. Whether the entry resolves to an ESM or a CJS file, `import` can handle it, and therefore so can the ESM loader. If the ESM loader can’t _resolve_ the entry, because it’s an unknown extension or extensionless or requires extension searching, it could try resolving it via `require.resolve` and if that works, use that result with `format: 'commonjs'`. The other advantage of always using the ESM loader is that `--loader` would work, and so custom loaders can expand what the ESM loader can resolve (such as to include `.mts` files).

This would be a breaking change for anyone launching an app like `node ./entry` where the entry point is `entry.mjs`, but that would arguably be a good thing, as it’s inconsistent for extension searching to apply to entry points and not to `import` specifiers. We could make an informative error message like we already do for `import` specifiers that lack extensions, when a file with a probable extension is nearby.

Potential problems with this approach involve breaking async tests, for anything that’s fully CommonJS now; would it matter if the async ESM loader was involved in startup? And we might have to special-case `.node` entry points, if that’s even supported (can you do `node entry.node`?) since you can’t yet `import` a `.node` file. I assume that JSON files can’t be used as entries so they shouldn’t matter.

**Reply by ljharb** [reply]
Would that break `node ./entry` where the entry point is `entry.js` (as CJS)? (probably one of the most common use cases in all of nodedom)

**Reply by cspotcode** [reply]
> This would be a breaking change for anyone launching an app like `node ./entry` where the entry point is `entry.mjs`

Does this work today?  Based on informal testing it does not; it only works for `.js`.  Additionally, the file extension is added *before* the URL is sent to `--loader` so it works with `--loader`, too.

<details>
<pre><code>
touch foo.js bar.mjs biff.cjs
echo 'export function resolve(url, context, defaultResolve) {
        console.log(url);
        return defaultResolve(url, context);
}' > loader.mjs
node --version # v17.4.0 in my testing
node ./foo # works
node ./bar # errors
node ./biff # errors
node --loader ./loader.mjs ./foo # Logs file:///home/ubuntu/dev/test/foo.js ; no error
</code></pre>
</details>

**Reply by bmeck** [reply]
.mjs and .cjs are not picked up normally currently, but .js and .node are

**Reply by GeoffreyBooth** [reply]
> .mjs and .cjs are not picked up normally currently, but .js and .node are

So it would break `node entry` where `entry` is `entry.js` in `"type": "module"`, but not `"type": "commonjs"`; and where `entry` is `entry.node`. But I think this is an acceptable breaking change to ship in Node 18, it’s a very easy fix for people and we can provide a guiding error message.

**Reply by bmeck** [reply]
it won't break in type:module since there is a weird (bug?) pre-resolve on entry points using the CJS loader even if actual resolve/load is done via ESM loader

**Reply by ljharb** [reply]
Since “type” shouldn’t affect entry point behavior, it shouldn’t break in either case. It doesn’t feel like an acceptable breaking change to me to invalidate the way people have been starting most node programs for over a decade.

**Reply by cspotcode** [reply]
So what I'm hearing is that this will *not* break anything, correct?

Building upon my previous informal testing:

```
echo '{"type": "module"}' > package.json
node ./foo # still works
node --loader ./loader.mjs ./foo # still logs the same file url with the js extension; still no error
```

**Reply by GeoffreyBooth** [reply]
> it won’t break in type:module since there is a weird (bug?) pre-resolve on entry points using the CJS loader even if actual resolve/load is done via ESM loader

This is what I’m proposing changing. Instead of resolving via the CJS loader first, we resolve via the ESM loader first and only try using the CJS resolution rules if ESM doesn’t resolve to anything. So `node ./entry` will continue to work _if_ `entry.js` is a CommonJS file and it’s in a `"type": "commonjs"` scope. But if it’s an ESM file in a `"type": "module"` scope, you need to write it as `node ./entry.js`. That’s it. Nothing changes for CommonJS entries; ESM entries need to add extensions. This fixes what’s arguably another bug, in that extension searching shouldn’t apply to ESM specifiers unless `--experimental-specifier-resolution=node` is enabled.

So @cspotcode in your example, you would need to change `./foo` to `./foo.js`. The extensionless form would no longer continue to work _because it’s a `.js` file in a `"type": "module"` scope._ If it were in a `"type": "commonjs"` scope, `node ./foo` would continue to work as before.

**Reply by cspotcode** [reply]
In your proposed behavior, we could probably also pass the result of the CJS resolve to the `--loader`.  Not sure yet if I love the idea, maybe there are gotchas, but:

#### Pros:

- we invoke the `--loader` for all entrypoints, no exceptions
- `node ./entry` within type:module doesn't break (though not sure how important this is to support)
- all other benefits of your approach still apply

#### Cons:

- `--loader` may be invoked twice for the entrypoint: once to fail resolving `./foo`, the second time given the `./foo.js` path from the CJS loader

#### Pseudocode

```typescript
const entrypoint = path.resolve(cwd, argv2);
try {
  const resolved = esmLoaderCodepath.resolve(fileUrl(entrypoint));
  // proceed with esm loader path, possible executing as CJS, possibly as ESM
  esmLoaderCodepath.load(resolved);
} catch(e: ResolutionFailedError) {
  const cjsResolved = cjsLoaderCodepath.resolve(entrypoint);
  // proceed with esm loader path, possible executing as CJS, possibly as ESM
  const resolved = esmLoaderCodepath.resolve(fileUrl(cjsResolved));
  esmLoaderCodepath.load(resolved);
}
```

### Re: breaking async tests

I haven't looked closely, but if node's *internal* ESM loader is secretly not doing anything async, it might be possible to invoke it sync in these cases.  Thus we'd be using the ESM path for all entrypoints, no exceptions.  But in the short term, when `--loader` is absent, it would be sync to avoid breaking tests.

**Reply by GeoffreyBooth** [reply]
> * `node ./entry` within type:module doesn’t break (though not sure how important this is to support)

You mean where `./entry` is `./entry.js`, and `entry.js` is an ESM file? In my proposal this _would_ break, in that you’d need to run it as `node ./entry.js`.

As for the cons, `--loader` doesn’t apply to CommonJS at all yet, so it running twice is well in the future.

Again, this refactor of how the entry point is parsed should be a follow-up PR, in my opinion, after a first PR that makes the flags irrelevant to how the entry is parsed. `node --loader ts-loader entry.mts` would break in between the two, but that’s okay IMO; during that time users could use a workaround of `node --loader ts-loader --input-type=module --eval 'import("./entry.mts")'` and it’s not the end of the world.

**Reply by cspotcode** [reply]
> In my proposal this would break, in that you’d need to run it as node ./entry.js.

Right, I'm saying that my augmentation to your proposal avoids that breakage.

> As for the cons, --loader doesn’t apply to CommonJS at all yet, so it running twice is well in the future.

I'm referring to the `resolve()` hook being invoked twice in my augmented proposal.  The resolve hook *does* already apply to CommonJS entrypoints today, and this is a good thing.

> `node --loader ts-loader entry.mts` would break in between the two, but that’s okay IMO

I'd rather implement the desired behavior in userspace loaders in the interim.  This gives users a good experience while waiting for node to implement the changes it wants.  Preventing userspace loaders from touching entrypoints prevents this, which isn't great for users.  Comes back to the correction space @guybedford mentioned earlier.

**Reply by guybedford** [reply]
The problem very much seems to be one of how to incrementally adjust the behaviour. Currently `node x` does a resolution of `x` that is outside of both the CJS and ESM resolvers, it's a sort of path resolution which is just extension adding and nothing else. This custom resolution logic then also has custom logic to determine when the ESM loader should apply, which can then in turn defer to ESM or CJS interpretation although forms an async execution.

I would still personally object to not being able to have loaders apply to certain types of entry points. If we could first land a PR that can incrementally say enable extensionless files, then we might be closer to something resembling parity. I'd be happy to help work on such a PR as well if we could determine what the exact consensus might be.

**Reply by GeoffreyBooth** [reply]
> I would still personally object to not being able to have loaders apply to certain types of entry points.

I don’t think this was suggested? Loaders can apply to any entry point, even extensionless or unknown-extension files; you would just need to specify the full filename including extension (if there is one), since custom loaders work via the ESM loader.

Currently loaders don’t really work with CommonJS files, so for now extensionless or unknown-extension files that are enabled via loader would need to be ESM. But extensionless CommonJS files already work, so I don’t see this as a problem.

**Reply by guybedford** [reply]
> so for now extensionless or unknown-extension files that are enabled via loader would need to be ESM

Are you suggesting that eg `"type": "module"` would apply to `.ts` files? I just don't want `node --loader ts-loader app.ts` to be any more complex than that, ever.

**Reply by GeoffreyBooth** [reply]
> Are you suggesting that eg "type": "module" would apply to .ts files?

No. `--loader` hooks into the ESM loader, and only works for files that are evaluated as a format other than `commonjs` (for now). The goal is to extend support to CommonJS too but that's a ways off.

**Reply by guybedford** [reply]
`--loader` must always be able to hook the resolver and treat any sensible Node.js entry point as ESM if necessary.

**Reply by GeoffreyBooth** [reply]
> `--loader` must always be able to hook the resolver and treat any sensible Node.js entry point as ESM if necessary.

Yes, in my proposal it would do that. That's why I'm suggesting ESM always tries to resolve first. The CommonJS resolution would only happen if the ESM `defaultResolve`, and any custom loaders that have been registered, all fail to resolve a runnable file.

**Reply by cspotcode** [reply]
> Currently loaders don’t really work with CommonJS files, so for now extensionless or unknown-extension files that are enabled via loader would need to be ESM.

Worth pointing out for anyone reading along: in practice, they do support CJS.  Today a `--loader` can also install a `require.extensions` hook [^1], so it can both classify any file as CJS *and* transform that file before execution.  Effectively we have full support today while we wait for node to implement a cleaner alternative.

My point is not that this is pretty, but it's *possible*.  This allows the ecosystem to experiment and for loader authors to keep users unblocked in the interim.

[^1]: I know this is because it's in the main thread and that may change

**Reply by GeoffreyBooth** [reply]
> can also install a require.extensions hook

Yes, this is true. When we were discussing loaders I was thinking of what could be achieved through the loaders API, not in a loader's code but using a different API. But yes, the “CommonJS loaders API” of `require.extensions` can fill some gaps.

**Comment by JakobJingleheimer** [comment]
# Proposal: ESM custom loader opt-in via export flag

Allow ESM custom loaders to opt into affecting entry-points. Ex

```mjs
// a-loader.mjs

export const useForEntryPoint = true;

export async function resolve() {}
export async function load() {}
```

This may introduce a cart-before-the-horse scenario (the custom loader might not have been itself loaded, so reading that export flag may _currently_ not be possible).

**Comment by JakobJingleheimer** [comment]
# Proposal: Create a “higher” class of custom loader (eg ambient-loader)

This class of custom loader would always be applied to entry-points (and may do other things too).

**Reply by JakobJingleheimer** [reply]
This sounds to me a potentially viable solution down the road but too large in scope for merely the current issue (and would thus take quite a while to hash out the technical design).

**Reply by bmeck** [reply]
I think this might be a different solution to a different problem, the size of this would be very large and have many potential side effects vs other potential solutions.

**Comment by JakobJingleheimer** [comment]
# Proposal: Treat extension-less file with "magic phrase" as JS

If and only if the file starts with `#!/usr/bin/env node\n`, parse as JavaScript text file (and follow regular rules).

**Reply by JakobJingleheimer** [reply]
I think this could work and aligns with historic precedent, but seems a little convoluted and cart-before-the-horse.

**Reply by ljharb** [reply]
This still doesn’t answer whether it’s ESM or CJS tho - overloading type module would mean that i couldn’t have .js ESM in the same folder as extensionless CJS, or vice versa.

**Reply by guybedford** [reply]
I would probably have a slight preference towards any magic string checks as applying specifically to the main `node x` entry point of the execution environment, as opposed to applying to any import.

Overloading the type is the same problem with wanting to have multiple `.js` meanings in the same app - one could create ` bin/package.json` with `"type": "commonjs"` set if wanting to have variation between files in the package per current process.

**Reply by JakobJingleheimer** [reply]
> This still doesn’t answer whether it’s ESM or CJS tho

@ljharb it would, per

> (and follow regular rules).

**Reply by JakobJingleheimer** [reply]
> I would probably have a slight preference towards any magic string checks as applying specifically to the main `node x` entry point of the execution environment, as opposed to applying to any import.

Yes, I think that was the intention of the original proposal 🙂

**Comment by guybedford** [comment]
# Definining "extensionless"

In relation to https://github.com/nodejs/node/discussions/41711#discussioncomment-2054908, another topic worth bringing up here is what is the definition of an extensionless file? There are a few that might apply:

* Is it a file without a `.` at all?
* Or is it a file that does not use a supported Node.js extension (ie not `.js` / `.json` / `.wasm` / `.node` / `.cjs` / `.mjs`)
* Or is it a file that does not use an extension from some MIME list or compound list of file extensions deemed to be the known extension list? If so, how would one construct such a list, and would it be a possible approach?

Under a specific definition of extensionless in the ESM loader it could then be possible for this definition to exhibit both backwards and forwards compatibility.

The backwards compatibility risk would remain that extensions in the reserved list would work in the CJS loader but not the ESM loader.

The forwards compatibility risk would remain that new extensions in future not in the reserved list would be breaking changes to support. Although perhaps this isn't such a huge risk since surely this is what MIME registries are designed to avoid in the first place.

**Reply by bmeck** [reply]
> Is it a file without a . at all?

That is what I believe this issue is about

> Or is it a file that does not use a supported Node.js extension (ie not .js / .json / .wasm / .node / .cjs / .mjs)

This seems to have an extension, but it is unknown to Node.js, so unknown extension might be a good term

> Or is it a file that does not use an extension from some MIME list or compound list of file extensions deemed to be the known extension list? If so, how would one construct such a list, and would it be a possible approach?

I'm not sure what this means

**Reply by ljharb** [reply]
if it has a dot, it’s not extensionless. Whether node knows about the extension has no bearing on whether it has one.

**Reply by guybedford** [reply]
Okay, that's definitely much easier to deal with then.

**Comment by cspotcode** [comment]
Coming back to this, it looks like the best options are either:
- Attempt the ESM loader for all entrypoints and fallback to the CommonJS loader when the ESM loader fails to return a URL or a format
- Teach the ESM loader's `default*` hooks to emulate CJS-style fallback when `node:entrypoint` condition / `parentURL` sentinel / `context.isMain` / internally-bound `isMain`

**Reply by GeoffreyBooth** [reply]
The first option, yes. The second option, I’m not sure what you’re referring to. I don’t think there should be any cases where the ESM loader is trying to “emulate” the CJS loader; it should just defer to the CJS loader itself if the ESM loader can’t resolve the entry point.

**Reply by cspotcode** [reply]
For that one, I'm thinking `defaultResolve()` can do the deferral if it's resolving an entrypoint.  So if a loader hook calls `defaultResolve()`, and it is resolving an entrypoint, and the CJS fallback is necessary, then the return value of `defaultResolve()` will be the result of the CJS fallback.

**Reply by bmeck** [reply]
I think the fallback semantics need to be worked out, they show up in at least 2 of the threads above but with different situations causing them. However, since 2 threads are proposing a similar solution perhaps we can gather around that.

**Reply by cspotcode** [reply]
Agreed; I was reluctant to create a new thread, but since the threads above are long enough that they're hard to follow, a new one makes sense to me.

To avoid confusion, it will help to define vocabulary for referring to node's internal ESM loader vs referring to `--loader` hooks.  The internal ESM loader is capable of classifying a file as CommonJS and causing execution of that file as CJS, so it is often possible to use the ESM loader in CJS contexts.  I worry that saying "ESM loader" might imply to some readers that CJS files cannot pass through that code path.

It feels like there are 3 things that might be thought of as "ESM Loader":

- the internal ESM loader, responsible for invoking --loader hooks if necessary, and triggering CJS execution if format is CJS
- user --loader hooks, which often rely on receiving a meaningful answer from `defaultResolve` or `defaultLoad`
- the implementations of `defaultResolve`, `defaultLoad`, and `defaultGetFormat` (`defaultGetFormat` might be confusing)

**Reply by GeoffreyBooth** [reply]
> I worry that saying “ESM loader” might imply to some readers

I don’t think casual users of Node will need to know the details. The docs will say something like “Node will try to resolve the entry point as an ES module; if it can be, it will be run as an ES module. If it can’t, Node will then try to resolve the entry point as a CommonJS module and run it as that.”

**Reply by cspotcode** [reply]
Indeed, I'm referring to how we proceed with our discussion here.
